src/mlpack/methods/cf/normalization/overall_mean_normalization.hpp

```diff
@@ -0,0 +1,111 @@
+/**
+ * @file overall_mean_normalization.hpp
+ * @author Wenhao Huang
+ *
+ * This class performs overall mean normalization on raw ratings. In another
+ * word, this class is used to remove global effect of overall mean.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_CF_NORMALIZATION_OVERALL_MEAN_NORMALIZATION_HPP
+#define MLPACK_METHODS_CF_NORMALIZATION_OVERALL_MEAN_NORMALIZATION_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace cf {
+
+/**
+ * This normalization class performs overall mean normalization on raw ratings.
+ */
+class OverallMeanNormalization
+{
+ public:
+  // Empty constructor.
+  OverallMeanNormalization() { }
+
+  /**
+   * Normalize the data by subtracting the mean of all existing ratings.
+   *
+   * @param data Input dataset in the form of coordinate list.
+   */
+  void Normalize(arma::mat& data)
+  {
+    mean = arma::mean(data.row(2));
+    data.row(2) -= mean;
+  }
+
+  /**
+   * Normalize the data by subtracting the mean of all existing ratings.
+   *
+   * @param cleanedData Sparse matrix data.
+   */
+  void Normalize(arma::sp_mat& cleanedData)
+  {
+    // Caculate mean of all non zero ratings.
+    if (cleanedData.n_nonzero != 0)
+      mean = arma::accu(cleanedData) / cleanedData.n_nonzero;
+    else
+      mean = 0;
+    // Subtract mean from all non zero ratings.
+    arma::sp_mat::iterator it = cleanedData.begin();
```

Comment:
We could put this block into the first condition and remove the else case, since ``cleanedData`` is going to be the same if ``cleanedData.n_nonzero == 0``.

Reply:
Yes we should put the block into the first condition like you suggested. But we still need the else case to set `mean = 0` because `mean` is a class member.